src/Datadog.Trace/Containers/ContainerInfo.cs

```diff
@@ -0,0 +1,118 @@
+        /// <remarks>
+        /// This method is used when passing the entire cgroup text for testing purposes.
+        /// When reading from file system, we only read one line at a time until we find a match,
+        /// which is usually the first line.
+        /// </remarks>
+        /// <param name="contents">The contents of a cgroup file.</param>
+        /// <returns>The container id if a match is found; otherwise, <c>null</c>.</returns>
+        public static string ParseCgroupText(string contents)
+        {
+            if (contents == null)
+            {
+                return null;
+            }
+
+            IEnumerable<string> lines = SplitLines(contents);
+            return ParseCgroupLines(lines);
+        }
+
+        /// <summary>
+        /// Used by <see cref="ParseCgroupText"/> to split the entire file contents
+        /// into individual lines to mock <see cref="File.ReadLines(string)"/>.
+        /// </summary>
+        /// <param name="contents">The multi-line string.</param>
+        /// <returns>An enumerable that returns each line from <paramref name="contents"/> when iterated.</returns>
+        public static IEnumerable<string> SplitLines(string contents)
+        {
+            if (contents == null)
+            {
+                yield break;
+            }
+
+            using (var reader = new StringReader(contents))
+            {
+                while (true)
+                {
+                    string line = reader.ReadLine();
+
+                    if (line == null)
+                    {
+                        yield break;
+                    }
+
+                    yield return line;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Uses regular expression to try to extract a container id from the specified string.
+        /// </summary>
+        /// <param name="lines">Lines of text from a cgroup file.</param>
+        /// <returns>The container id if found; otherwise, <c>null</c>.</returns>
+        public static string ParseCgroupLines(IEnumerable<string> lines)
+        {
+            return lines.Select(ParseCgroupLine)
+                        .FirstOrDefault(id => !string.IsNullOrWhiteSpace(id));
+        }
+
+        /// <summary>
+        /// Uses regular expression to try to extract a container id from the specified string.
+        /// </summary>
+        /// <param name="line">A single line from a cgroup file.</param>
+        /// <returns>The container id if found; otherwise, <c>null</c>.</returns>
+        public static string ParseCgroupLine(string line)
+        {
+            var lineMatch = Regex.Match(line, ContainerIdRegex);
+
+            return lineMatch.Success
+                       ? lineMatch.Groups[1].Value
+                       : null;
+        }
+
+        private static string GetContainerIdInternal()
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) && File.Exists(ControlGroupsFilePath))
```

Comment:
`File.Exists` returns `false` if the caller doesn't have sufficient permissions to read the file. Do we ever foresee that happening? Probably not, right?

Reply:
My concern (which may be unwarranted) is that just relying on the return value of `File.Exists` is that we can't differentiate between 1) a machine that is not a container and 2) a machine that for some reason doesn't have permissions to read that file. I'm just not knowledgeable enough about containers to know if case 2 can ever happen.

Reply:
Good catch. I added exception handling and logging in d78a4cc1f7a92714e4c4aabef45c5dd7a086c4f5.

Reply:
Currently the logic is:
- if the OS is Linux
- and if the file exists
- and if we can read the file
- and if we can extract a containerId from any line in the file (see regex)

...then we return the containerId. Otherwise, we return null.

Reply:
To be clear, `/proc/self/cgroup` is not a real file. There's a virtual filesystem mounted on `/proc` with information about running processes. `/proc/self` is a dynamic symlink to `/proc/{process_id}` which contains information about the current process.

http://manpages.ubuntu.com/manpages/xenial/man5/proc.5.html